testsuite/tests/typing-jkind-bounds/subsumption/quality.ml

```diff
@@ -0,0 +1,286 @@
+         because of the definition of u at line 3, characters 0-6.
+       But the kind of type "u" must be a subkind of value mod global with a
+         because of the definition of t at line 4, characters 0-36.
+|}]
+
+module F (M : sig type t end) = struct
+  module type S = sig
+    type t : value mod global with M.t
+  end
+  type t
+  module type T = S with type t = t
+end
+[%%expect {|
+Line 6, characters 25-35:
+6 |   module type T = S with type t = t
+                             ^^^^^^^^^^
+Error: The kind of type "t" is value
+         because of the definition of t at line 5, characters 2-8.
+       But the kind of type "t" must be a subkind of value mod global with M.t
+         because of the definition of t at line 3, characters 4-38.
+|}]
+
+module type S = sig
+  type t
+end
+
+type a
+module M : S with type t = a = struct
+  type t = a
+end
+
+module _ : sig
+  type t : value mod portable uncontended with M.t
+end = struct
+  type t : value mod portable
+end
+[%%expect {|
+module type S = sig type t end
+type a
+module M : sig type t = a end
+Lines 12-14, characters 6-3:
+12 | ......struct
+13 |   type t : value mod portable
+14 | end
+Error: Signature mismatch:
+       Modules do not match:
+         sig type t : value mod portable end
+       is not included in
+         sig type t : value mod uncontended portable with M.t end
+       Type declarations do not match:
+         type t : value mod portable
+       is not included in
+         type t : value mod uncontended portable with M.t
+       The kind of the first is value mod portable
+         because of the definition of t at line 13, characters 2-29.
+       But the kind of the first must be a subkind of
+         value mod uncontended portable with M.t
+         because of the definition of t at line 11, characters 2-50.
+|}]
+
+module type S = sig
+  type t
+  type u = t
+end
+
+type a
+module M : S with type t := a = struct
+  type t = a
+  type u = a
+end
+
+module M : sig
+  type t : value mod portable uncontended with M.u
+end = struct
+  type t : value mod portable
+end
+[%%expect {|
+module type S = sig type t type u = t end
+type a
+module M : sig type u = a end
+Lines 14-16, characters 6-3:
+14 | ......struct
+15 |   type t : value mod portable
+16 | end
+Error: Signature mismatch:
+       Modules do not match:
+         sig type t : value mod portable end
+       is not included in
+         sig type t : value mod uncontended portable with M.u end
+       Type declarations do not match:
+         type t : value mod portable
+       is not included in
+         type t : value mod uncontended portable with M.u
+       The kind of the first is value mod portable
+         because of the definition of t at line 15, characters 2-29.
+       But the kind of the first must be a subkind of
+         value mod uncontended portable with M.u
+         because of the definition of t at line 13, characters 2-50.
+|}]
+
+module M : sig
+  type a = [`a of string | `b]
```

Comment:
I actually think it would be fine to accept this: there is no row variable here (no `<` or `>` in the poly-variant). But I would expect to reject for now. Maybe leave a comment saying it would be fine to accept -- and then also test the cases with row variables.

Reply:
done